sdk/dotnet/Pulumi/Resources/Resource.cs

```diff
@@ -0,0 +1,361 @@
+        /// can be a parent of other resources, we don't want to ever depend on those child
+        /// resource.  If we do, it's simple to end up in a situation where we end up depending on a
+        /// child resource that has a data cycle dependency due to the data passed into it. An
+        /// example of how this would be bad is:
+        /// 
+        /// <c>
+        ///     var c1 = new CustomResource("c1");
+        ///     var c2 = new CustomResource("c2", { parentId = c1.id }, { parent = c1 });
+        ///     var c3 = new CustomResource("c3", { parentId = c1.id }, { parent = c1 });
+        /// </c>
+        /// 
+        /// The problem here is that 'c2' has a data dependency on 'c1'.  If it tries to wait on
+        /// 'c1' it will walk to the children and wait on them.This will mean it will wait on 'c3'.
+        /// But 'c3' will be waiting in the same manner on 'c2', and a cycle forms. This normally
+        /// does not happen with ComponentResources as they do not have any data flowing into
+        /// them.The only way you would be able to have a problem is if you had this sort of coding
+        /// pattern:
+        /// 
+        /// <c>
+        ///     var c1 = new ComponentResource("c1");
+        ///     var c2 = new CustomResource("c2", { parentId = c1.urn }, { parent: c1 });
+        ///     var c3 = new CustomResource("c3", { parentId = c1.urn }, { parent: c1 });
+        /// </c>
+        /// 
+        /// However, this would be pretty nonsensical as there is zero need for a custom resource to
+        /// ever need to reference the urn of a component resource.  So it's acceptable if that sort
+        /// of pattern failed in practice.
+        /// </summary>
+        internal readonly HashSet<Resource> ChildResources = new HashSet<Resource>();
+
+        /// <summary>
+        /// Urn is the stable logical URN used to distinctly address a resource, both before and
+        /// after deployments.
+        /// </summary>
+
+        internal readonly UrnOutputCompletionSource _urn;
+        public Output<Urn> Urn => _urn.Output;
+
+        /// <summary>
+        /// When set to true, protect ensures this resource cannot be deleted.
+        /// </summary>
+        private readonly bool _protect;
+
+        /// <summary>
+        /// A collection of transformations to apply as part of resource registration.
+        /// </summary>
+        private readonly ImmutableArray<ResourceTransformation> _transformations;
+
+        /// <summary>
+        /// A list of aliases applied to this resource.
+        /// </summary>
+        internal readonly ImmutableArray<Input<Urn>> _aliases;
+
+        /// <summary>
+        /// The type assigned to the resource at construction.
+        /// </summary>
+        public string Type { get; }
+
+        /// <summary>
+        /// The name assigned to the resource at construction.
+        /// </summary>
+        public string Name { get; }

```

Comment:
This `Name` currently contains a logical name that I pass as the first argument to the constructor (e.g. `rg`).
I can't define an output `Name` property on a specific resource (e.g. `ResourceGroup`) because it would shadow the parent property. So, for now, I had to call it obviously ugly `Name1`. What's the solution here?

``` cs
    public class ResourceGroup : CustomResource
    {
        [ResourceField("name")]
        private readonly StringOutputCompletionSource _name;
        public Output<string> Name1 => _name.Output;
```

Reply:
> I can't define an output Name property on a specific resource (e.g. ResourceGroup) because it would shadow the parent property. 

Shadowing is legal in .net.  So that should be ok.  

I can make these properties internal , but it made some stuff uglier because we end up having to pass around a Resource+Name+Type (which just seems redundant).

Reply:
Fixed by making the equiv resource entrypoints into methods.